src/coreclr/jit/morph.cpp

```diff
@@ -11053,6 +11053,69 @@ GenTree* Compiler::fgMorphCommutative(GenTreeOp* tree)
     return op1;
 }
 
+//------------------------------------------------------------------------------
+// fgMorphCastedBitwiseOp : Try to simplify "(T)x op (T)y" to "(T)(x op y)".
+//
+// Arguments:
+//     tree - node to fold
+//
+// Return Value:
+//     A folded GenTree* instance, or nullptr if it couldn't be folded
+GenTree* Compiler::fgMorphCastedBitwiseOp(GenTreeOp* tree)
+{
+    assert(varTypeIsIntegralOrI(tree));
+    assert(tree->OperIs(GT_OR, GT_AND, GT_XOR));
+
+    GenTree* op1 = tree->gtGetOp1();
+    GenTree* op2 = tree->gtGetOp2();
+    genTreeOps oper = tree->OperGet();
+
+    // see whether both ops are casts, with matching to and from types.
+    if (op1->OperIs(GT_CAST) && op2->OperIs(GT_CAST))
+    {
+        // bail if either operand is a checked cast
+        if (op1->gtOverflowEx() || op2->gtOverflowEx())
+        {
+            return nullptr;
+        }
+
+        var_types fromType = op1->AsCast()->CastOp()->TypeGet();
+        var_types toType = op1->AsCast()->gtCastType;
+        bool isUnsigned = op1->IsUnsigned();
+
+        if ((op2->CastFromType() != fromType) || (op2->CastToType() != toType) || (op2->IsUnsigned() != isUnsigned))
+        {
+            return nullptr;
+        }
+
+        // Reuse gentree nodes:
+        //
+        //     tree             op1
+        //     /   \             |
+        //   op1   op2   ==>   tree
+        //    |     |          /   \
+        //    x     y         x     y
+        //
+        // (op2 becomes garbage)
+
+        tree->gtOp1 = op1->AsCast()->CastOp();
+        tree->gtOp2 = op2->AsCast()->CastOp();
+        tree->gtType = fromType;
+
+        op1->gtType = genActualType(toType);
+        op1->AsCast()->gtOp1 = tree;
```

Comment:
```suggestion
        op1->AsCast()->gtOp1 = tree;
        op1->SetAllEffectsFlags(tree);
```
Now that we're reusing the node, we need to reset the side effect flags (should fix the CI failures).